internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,385 @@
+		)
+		basicLimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+		if (limits.memoryLimitInMB << 20) > math.MaxUint64 {
+			return fmt.Errorf("memory limit value of `%d` is invalid", limits.memoryLimitInMB)
+		}
+		eliInfo.JobMemoryLimit = (limits.memoryLimitInMB << 20)
+		eliInfo.BasicLimitInformation.LimitFlags = basicLimitFlags
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.cpuRate != 0 || limits.cpuWeight != 0 || limits.cpuCount != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.cpuCount != 0 {
+			numProcessors := uint32(processorinfo.ProcessorCount())
+			if limits.cpuCount > numProcessors {
+				limits.cpuCount = numProcessors
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = calculateJobCPURate(limits.cpuCount, numProcessors)
+		} else if limits.cpuWeight != 0 {
+			if limits.cpuWeight < processorWeightMin || limits.cpuWeight > processorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.cpuWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = calculateJobCPUWeight(limits.cpuWeight)
+		} else if limits.cpuRate != 0 {
+			if limits.cpuRate < processorLimitMin || limits.cpuRate > processorLimitMax {
+				return fmt.Errorf("processor rate of `%d` is invalid", limits.cpuRate)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = limits.cpuRate
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectCpuRateControlInformation, uintptr(unsafe.Pointer(&cpuInfo)), uint32(unsafe.Sizeof(cpuInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.maxBandwidth != 0 || limits.maxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.maxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.maxBandwidth
+		}
+		if limits.maxIops != 0 {
+			ioInfo.MaxIops = limits.maxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// Create creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func Create(name string) (*JobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
```

Comment:
> StringToUTF16Ptr is deprecated. Use UTF16PtrFromString instead.

Reply:
Also, this seems like calling `Create` with an empty name will result in `CreateJobObject` being passed an empty string, rather than `NULL`, which is what is expected to indicate to create an unnamed job object. Have you tested with `Create("")`?

Reply:
This succeeds, although we should just check if `name == ""` and pass in nil

Reply:
Agreed